File: src/fragment/SlotFragment.ts
import { HTMLText, Text } from 'pixi.js';
import Fragment, { FragmentType, fragmentTypes, register } from './Fragment';
import { StructEndecBuilder, PrimitiveEndecs, Optional } from 'KEndec';
import { Either, EitherEndec, Identifier } from '~/util';
import { ALWAYS_READABLE_BLOCK_POS, safeOptionalOf, UUID, VectorI } from '~/endecTomfoolery';

type Vector = { x: number; y: number; z: number };

export default class SlotFragment extends Fragment {
    slot: number;
    source: Optional<Either<VectorI, any /*UUID*/>>;

    constructor(slot: number, source: Optional<Either<VectorI, any /*UUID*/>>) {
        super();

        this.slot = slot;
        this.source = source;
    }

    override asFormattedText(): HTMLText {
        const entityColor = fragmentTypes.get(new Identifier('trickster', 'entity'))!.color;
        const numColor = fragmentTypes.get(new Identifier('trickster', 'number'))!.color;
        const coloredInt = (value: number) =>
            `<span style="color: #${numColor.toString(16)}">${value}</span>`;

        var sourceText;
        if (this.source.isEmpty()) {
            sourceText = `<span style="color: #${entityColor.toString(16)}">Caster</span>`;
        } else if (this.source.get().left().isPresent()) {
            const source = this.source.get().left().get();
            sourceText = `(${coloredInt(source.x)}, ${coloredInt(source.y)}, ${coloredInt(source.z)})`;
        } else if (this.source.get().right().isPresent()) {
            const source = this.source.get().right().get();
            sourceText = `<span style="color: #${entityColor.toString(16)}">${source}</span>`;
        }

        const slotText = coloredInt(this.slot);

        return new HTMLText({
            text: `<span style="color: #${this.type().color.toString(
                16
            )}">slot ${slotText} at ${sourceText}</span>`,
        });
    }

    override toString(): string {
        var sourceText = 'Caster';
        if (this.source.isPresent()) {
            if (this.source.get().right().isPresent()) {
                sourceText = this.source.get().right().get();
            } else if (this.source.get().left().isPresent()) {
                const source = this.source.get().right().get();

                sourceText = `(${source.x}, ${source.y}, ${source.z})`;
            }
        }

        return `slot ${this.slot} at ${sourceText}`;
    }

    override type(): FragmentType<SlotFragment> {
        return SLOT;
    }
}

const SLOT = register(
    'slot',
    0x77aaee,
    StructEndecBuilder.of2(
        PrimitiveEndecs.INT.fieldOf('slot', (framgent: SlotFragment) => framgent.slot),
        safeOptionalOf(new EitherEndec(ALWAYS_READABLE_BLOCK_POS, UUID, true)).fieldOf(
            'source',
            (fragment: SlotFragment) => fragment.source
        ),
        (slot, source) => new SlotFragment(slot, source)
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/SlotFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import Either
import EitherEndec
import Identifier
import com.benasher44.uuid.Uuid
import io.kvision.utils.obj
import me.maplesyrum.tricksterstudio.endec.ALWAYS_READABLE_BLOCK_POS
import me.maplesyrum.tricksterstudio.endec.UUID
import me.maplesyrum.tricksterstudio.endec.Vector
import me.maplesyrum.tricksterstudio.endec.VectorI
import me.maplesyrum.tricksterstudio.endec.safeOptionalOf
import me.maplesyrum.tricksterstudio.external.pixi.HTMLText
import tree.maple.kendec.PrimitiveEndecs
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder
import tree.maple.kendec.util.Optional


class SlotFragment(val slot: Int, val source: Optional<Either<VectorI, Uuid>>) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.SLOT
    }

    override fun asFormattedText(): HTMLText {
        val entityColor = fragmentTypes[Identifier("trickster", "entity")]!!.color.get()
        val numColor = fragmentTypes[Identifier("trickster", "number")]!!.color.get()
        fun coloredInt(value: Int) = "<span style=\"color: #${numColor.toString(16)}\">$value</span>"

        val sourceText = when {
            source.isEmpty() -> "<span style=\"color: #${entityColor.toString(16)}\">Caster</span>"
            source.get().left().isPresent() -> {
                val src = source.get().left().get()
                "(${coloredInt(src.x)}, ${coloredInt(src.y)}, ${coloredInt(src.z)})"
            }

            source.get().right().isPresent() -> {
                val src = source.get().right().get()
                "<span style=\"color: #${entityColor.toInt().toString(16)}\">$src</span>"
            }

            else -> ""
        }

        val slotText = coloredInt(slot)

        return HTMLText(
            obj {
                text = "<span style=\"color: #${type().color.map(::toInt).orElse(0xaaaaaa).toString(16)}\">slot $slotText at $sourceText</span>"
            }
        )
    }

    override fun toString(): String {
        var sourceText = "Caster"
        if (source.isPresent()) {
            if (source.get().right().isPresent()) {
                sourceText = source.get().right().get().toString()
            } else if (source.get().left().isPresent()) {
                val src = source.get().left().get()
                sourceText = "(${src.x}, ${src.y}, ${src.z})"
            }
        }
        return "slot $slot at $sourceText"
    }

    companion object {
        val ENDEC: StructEndec<SlotFragment> = StructEndecBuilder.of(
            PrimitiveEndecs.INT.fieldOf("slot", SlotFragment::slot),
            safeOptionalOf(
                EitherEndec(
                    ALWAYS_READABLE_BLOCK_POS,
                    UUID,
                    true
                )
            ).fieldOf("source", SlotFragment::source),
            { slot: Int, source: Optional<Either<VectorI, Uuid>> -> SlotFragment(slot, source) }
        )
    }
}
